### PR TITLE
SONARPY-583 Rule S5717: Function parameters' default values should not be modified or assigned

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(FUNCDEF, ctx -> {
+      FunctionDef functionDef = (FunctionDef) ctx.syntaxNode();
+      // avoid raising issues on nested function, it may have been done on purpose
+      if (TreeUtils.firstAncestorOfKind(functionDef, FUNCDEF) != null) {
+        return;
+      }
+      for (Parameter parameter : nonTupleParameters(functionDef)) {
+
```
Comment (gyula-sallai-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: Nitpick: I would remove this empty line.

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+import static org.sonar.plugins.python.api.tree.Tree.Kind.NAME;
+import static org.sonar.plugins.python.api.tree.Tree.Kind.QUALIFIED_EXPR;
+import static org.sonar.plugins.python.api.tree.Tree.Kind.SUBSCRIPTION;
+import static org.sonar.python.tree.TreeUtils.getSymbolFromTree;
+import static org.sonar.python.tree.TreeUtils.nonTupleParameters;
+
+@Rule(key = "S5717")
+public class ModifiedParameterValueCheck extends PythonSubscriptionCheck {
+
+  private static final Map<String, Set<String>> MUTATING_METHODS = new HashMap<>();
```
Comment (gyula-sallai-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: All the other static members are declared in the line preceding the `static {}` block in which they are initialized, I would suggest to do the same for this one as well and move the declaration to line 92.

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    if (typeMutatingMethods == null) {
+      return Collections.emptyList();
+    }
+    return paramSymbol.get().usages().stream().filter(usage -> {
+      Tree parent = usage.tree().parent();
+      if (parent.is(QUALIFIED_EXPR)) {
+        QualifiedExpression qualifiedExpression = (QualifiedExpression) parent;
+        return  getSymbolFromTree(qualifiedExpression.qualifier()).equals(paramSymbol) && isMutatingMethod(typeMutatingMethods, qualifiedExpression.name().name());
+      }
+      return isUsedInDelStatement(usage.tree()) ||
```
Comment (gyula-sallai-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: I find this block of code hard to read due to all the returns and tight formatting. I think it should be a little bit more spacious, I would introduce some blank lines between the control structures.

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    String defaultValueType = defaultValueType(defaultValue);
+    Set<String> typeMutatingMethods = MUTATING_METHODS.get(defaultValueType);
+    if (typeMutatingMethods == null) {
+      return Collections.emptyList();
+    }
+    return paramSymbol.get().usages().stream().filter(usage -> {
+      Tree parent = usage.tree().parent();
+      if (parent.is(QUALIFIED_EXPR)) {
+        QualifiedExpression qualifiedExpression = (QualifiedExpression) parent;
+        return  getSymbolFromTree(qualifiedExpression.qualifier()).equals(paramSymbol) && isMutatingMethod(typeMutatingMethods, qualifiedExpression.name().name());
```
Comment (gyula-sallai-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: Superfluous whitespace.

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+      .filter(tree -> isUsedInLhsOfAssignment(tree, expression -> isQualifier(tree, expression)))
+      .collect(Collectors.toList());
+  }
+
+  private static boolean isUsedInLhsOfCompoundAssignment(Tree tree) {
+    CompoundAssignmentStatement compoundAssignmentStatement = ((CompoundAssignmentStatement) TreeUtils.firstAncestorOfKind(tree, COMPOUND_ASSIGNMENT));
+    if (compoundAssignmentStatement == null) {
+      return false;
+    }
+    return isAccessingExpression(compoundAssignmentStatement.lhsExpression(), tree);
```
Comment (gyula-sallai-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: Nitpick: I would simplify this to `return compoundAssignmentStatement != null && isAccessingExpression(...)`

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+  }
+
+  private static boolean isGetItemOnDefaultDict(@Nullable String defaultValueType, Tree tree) {
+    if (!"collections.defaultdict".equals(defaultValueType)) {
+      return false;
+    }
+    return isObjectOfSubscription(tree, tree.parent());
```
Comment (gyula-sallai-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: Nitpick: same as above.

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+   * Detects if shape of tree is equal to
+   * - expression[SOMETHING]
+   * - expression.SOMETHING
+   */
+  private static boolean isAccessingExpression(Expression expression, Tree tree) {
+    return isObjectOfSubscription(tree, expression) || isQualifier(tree, expression);
+  }
+
+  /**
+   *
```
Comment (gyula-sallai-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: Unneeded line.

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java (hunk 2 above)
Comment (andrea-guarino-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: Good point, Thanks!

File: python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java (hunk 3 above)
Comment (andrea-guarino-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/ModifiedParameterValueCheck.java]: I decided to extract a function to make it more readable. Thanks!

**Pull request merged**